tools/fix_bss.py

```diff
@@ -0,0 +1,649 @@
+# Note that we may not be able to match all variables if a variable is not referenced by any pointer.
+def determine_base_bss_ordering(
+    build_bss_symbols: list[BssSymbol], pointers: list[Pointer]
+) -> list[BssSymbol]:
+    # Assume that the lowest address is the start of the BSS section
+    base_section_start = min(p.base_value for p in pointers)
+    build_section_start = min(p.build_value for p in pointers)
+
+    found_symbols: dict[str, BssSymbol] = {}
+    for p in pointers:
+        base_offset = p.base_value - base_section_start
+        build_offset = p.build_value - build_section_start
+
+        new_symbol = None
+        new_offset = 0
+        for symbol in build_bss_symbols:
+            if (
+                symbol.offset <= build_offset
+                and build_offset < symbol.offset + symbol.size
+            ):
+                new_symbol = symbol
+                new_offset = base_offset - (build_offset - symbol.offset)
+                break
+
+        if new_symbol is None:
+            if p.addend > 0:
+                addend_str = f"+0x{p.addend:X}"
+            elif p.addend < 0:
+                addend_str = f"-0x{-p.addend:X}"
+            else:
+                addend_str = ""
+            fail(
+                f"Error: could not find BSS symbol for pointer {p.name}{addend_str} "
+                f"(base address 0x{p.base_value:08X}, build address 0x{p.build_value:08X})"
+            )
+            return []
+
+        if new_symbol.name in found_symbols:
+            # Sanity check that offsets agree
+            existing_offset = found_symbols[new_symbol.name].offset
+            if new_offset != existing_offset:
+                fail(
+                    f"Error: BSS symbol {new_symbol.name} found at multiple offsets in baserom "
+                    f"(0x{existing_offset:04X} and 0x{new_offset:04X})"
+                )
+        else:
+            found_symbols[new_symbol.name] = BssSymbol(
+                new_symbol.name, new_offset, new_symbol.size, new_symbol.align
+            )
+
+    return list(sorted(found_symbols.values(), key=lambda symbol: symbol.offset))
+
+
+# Determine a new set of increment_block_number pragmas that will fix the BSS ordering.
+def solve_bss_ordering(
+    pragmas: list[Pragma],
+    bss_variables: list[BssVariable],
+    base_bss_symbols: list[BssSymbol],
+) -> list[Pragma]:
+    base_symbols_by_name = {symbol.name: symbol for symbol in base_bss_symbols}
+
+    # Our "algorithm" just tries all possible combinations of increment_block_number amounts.
+    # This can get really slow, so we first try multiples of 10 only, since the exact
+    # amount often doesn't matter much and we can find a solution faster.
+    fast_candidates = itertools.product(
+        [i for i in range(256) if i % 10 == 0], repeat=len(pragmas)
+    )
+    slow_candidates = itertools.product(range(256), repeat=len(pragmas))
+    all_candidates = itertools.chain(fast_candidates, slow_candidates)
```

Comment:
tbh I used powers of 10 because I think it looks nicer in the source files

Reply:
Here's a suggestion for the candidates to test for in order

<details>

```py
def nu2(v: int):
    """
    Returns 2-adic valuation of n for 0<n<256
    Returns 8 for n=0
    https://en.wikipedia.org/wiki/P-adic_valuation
    """
    if v == 0:
        # 0 = 256 mod 256
        return 8
    for nu in reversed(range(8)):
        if v % (2**nu) == 0:
            return nu
    assert False


seq_dim1 = [
    (v, nu2(v)) for v in (int("".join(reversed(f"{v:08b}")), 2) for v in range(256))
]


def gen_seq_product_impl(n: int):
    if n == 1:
        for v, nu in seq_dim1:
            yield (v,), nu
    else:
        # each iteration of this loop is for yielding all the products where the
        # lowest 2-adic value among the product's values is exactly min_nu_outer
        for min_nu_outer in reversed(range(9)):  # 8..0
            # this loop is for doing the cross product of the 1D sequence with the
            # (n-1) dimensions sequence
            for v, nu in seq_dim1:
                if nu < min_nu_outer:
                    # (as for other dimensions,) seq_dim1 is ordered by decreasing nu
                    # we can exit this loop since adding v to any (n-1)D product would
                    # make that product with a lowest 2-adic value < min_nu_outer
                    break
                # the next two elif are based on the fact that
                # [gen_seq(n) where lowest 2-adic value is min_nu_outer]
                #  =   [gen_seq(1) where 2-adic value is > min_nu_outer]x[gen_seq(n-1) where lowest 2-adic value is = min_nu_outer] (first elif)
                #    U [gen_seq(1) where 2-adic value is = min_nu_outer]x[gen_seq(n-1) where lowest 2-adic value is < min_nu_outer] (second elif)
                elif nu > min_nu_outer:
                    for prevdim_prod, prevdim_prod_lownu in gen_seq_product_impl(n - 1):
                        if prevdim_prod_lownu < min_nu_outer:
                            break
                        if prevdim_prod_lownu != min_nu_outer:
                            # skip prevdim_prod_lownu > min_nu_outer, but keep looping until finding the prevdim_prod_lownu == min_nu_outer ones
                            # this is a performance problem, ideally we would add an argument to gen_seq_product_impl for asking specific lowest 2-adic values
                            continue
                        yield (v, *prevdim_prod), min_nu_outer
                elif nu == min_nu_outer:
                    for prevdim_prod, prevdim_prod_lownu in gen_seq_product_impl(n - 1):
                        if prevdim_prod_lownu < min_nu_outer:
                            break
                        yield (v, *prevdim_prod), min_nu_outer
                else:
                    assert False


def gen_seq_product(n: int):
    """
    Yields elements of the cartesian product [0;255]^n,
    ordered by descending min(2-adic value of elements in the product)
    """
    for prod, lownu in gen_seq_product_impl(n):
        yield prod


for p in gen_seq_product(3):
    print(p)
    input()
```

</details>

it implements returning a cartesian product like here but in an order where first are the products with multiples of 256 (that's `(0,0,0)`), then 128, then 64, then 32, etc.

Here's the start of the sequence in 3D (3 pragmas):

<details>

```
(0, 0, 0)
(0, 0, 128)
(0, 128, 0)
(0, 128, 128)
(128, 0, 0)
(128, 0, 128)
(128, 128, 0)
(128, 128, 128)
(0, 0, 64)
(0, 0, 192)
(0, 128, 64)
(0, 128, 192)
(0, 64, 0)
(0, 64, 128)
```

</details>

Reply:
Huh really interesting. I tried to golf this a little using your suggestion:

```python
# Generate a sequence of integers with a 2-adic value of exactly nu
def gen_seq(nu: int):
    if nu == 8:
        yield 0
    else:
        for i in range(1 << (7 - nu)):
            yield (2*i + 1) * (1 << nu)


# n-ary product, where the lowest 2-adic value among the tuples's values is exactly min_nu
def gen_seq_product_impl(n: int, min_nu: int):
    if n == 1:
        for n in gen_seq(min_nu):
            yield (n,)
    else:
        # (a, *b) has min 2-adic value = min_nu if and only if either:
        #   a has 2-adic value >  min_nu and b has min 2-adic value == min_nu
        #   a has 2-adic value == min_nu and b has min 2-adic value >= min_nu
        for min_nu_inner in reversed(range(min_nu + 1, 9)):
            for a in gen_seq(min_nu_inner):
                for b in gen_seq_product_impl(n - 1, min_nu):
                    yield (a, *b)
        for a in gen_seq(min_nu):
            for min_nu_inner in reversed(range(min_nu, 9)):
                for b in gen_seq_product_impl(n - 1, min_nu_inner):
                    yield (a, *b)


# n-ary product, where the lowest 2-adic value among the tuples's values is exactly min_nu
def gen_seq_product(n: int):
    for nu in reversed(range(9)):
        yield from gen_seq_product_impl(n, nu)
```

Reply:
Nice, cool rewrite
throw caching on top of gen_seq and gen_seq_product_impl ( https://docs.python.org/3/library/functools.html#functools.cache ) and it's probably perfect

Reply:
We could probably cache them as lists but meh
Alright, pushed the change